### ISSUE TITLE
Finish production submission of Tabula Muris

### ISSUE THREAD

**Issue by malloryfreeberg** [HumanCellAtlas/ingest-central#343]
The Tabula Muris dataset in production is stuck at having exported 99,495/99,840 bundles. Remaining bundles need to be exported and the submission process go to completion.

submission ID: `5c51805d5fc0000007998f72`
project UUID: `f306435b-4e60-4a79-83a1-159bda5a2c79`

**Comment by malloryfreeberg** [comment]
@justincc

**Comment by rdgoite** [comment]
I did some investigation on this issue and found some notable points. Looking at the logs through Kibana, there seems to be at least 2 sources of the problem, one is with the upload service returning a HTTP 404 when Ingest called patch, and another is with the data store responding with HTTP 500 when Ingest tried to upload some bundle files (i.e. metadata JSON files included in a bundle) to the data store. 

Based on the logs I gathered through Kibana, there are a total of 103 bundle files that failed with 404 response from upload service, while there are 45 that failed with "too many 500 responses" from the data store. Combining all, there are only 134 unique files that encountered either of the errors. I've compiled all [relevant logging information I could find](https://drive.google.com/file/d/1tKnzT0EEhH3_40WnOiJ78Qi_692O-lI-/view?usp=sharing).

**Comment by malloryfreeberg** [comment]
> data store responding with HTTP 500 when Ingest tried to upload some bundle files

wasn't this issue supposedly fixed by enabling retries?

**Comment by rdgoite** [comment]
The system was complaining that there were too many retries triggered by HTTP 500 responses.

**Comment by justincc** [comment]
Rodrey is going to

a) Notify upload and datastore of the bundles where we think they may have had a problem
b) Write a proposal for improving our bundle manifest that we can discuss in the dev team. This is to make it easier to track down the bundles where we currently have no information, rather than having to do this by a painful manual process.

**Comment by parthshahva** [comment]
a) is a known problem that had already been discussed with Rolando. It was known that a certain # of bundles would have to be retried for this submission

**Comment by justincc** [comment]
@parthshahva What proportion of retries are acceptable?

**Comment by parthshahva** [comment]
This is not something we want to be doing ongoing. This is a known bug thats open and being debugged/addressed

**Comment by justincc** [comment]
@rdgoite @rolando-ebi Are we not retrying these upload calls at the moment?

**Comment by justincc** [comment]
I have created #349 to cover this work. @rdgoite in the meantime can we retry these particular failures?

**Comment by aaclan-ebi** [comment]
We're doing retries. We're using retry policy in the session. It should be retrying within 20mins with backoff, https://github.com/HumanCellAtlas/ingest-client/blob/38f8ff84378680134f7d3c55483310f4b188c44e/ingest/api/stagingapi.py#L35

Could we confirm in the logs if those error didn't get successful eventually?

**Comment by rdgoite** [comment]
We have log entries like this:
```
 2019-02-28 17:37:54,675  - ingest.exporter.ingestexportservice - INFO in ingestexportservice.py:475 verify_files(): File bed639fc-def9-480e-af1e-63ce36bcc9c6/2019-01-30T211349.121000Z with name process_0.json is successfully copied!
...skipping...
  File "/app/src/hca-ingest/ingest/exporter/ingestexportservice.py", line 408, in upload_metadata_files
    uploaded_file = self.upload_file(submission_uuid, filename, content, content_type)
  File "/app/src/hca-ingest/ingest/exporter/ingestexportservice.py", line 565, in upload_file
    raise e
  File "/app/src/hca-ingest/ingest/exporter/ingestexportservice.py", line 557, in upload_file
    file_description = self.staging_api.stageFile(submission_uuid, filename, content, content_type)
  File "/app/src/hca-ingest/ingest/api/stagingapi.py", line 98, in stageFile
    r.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://upload.data.humancellatlas.org/v1/area/d3d2f98c-9b8e-4ddb-8534-2d437aa15379/process_ecc8f8b8-b655-48f5-a120-1bd282411407.json
```

Admittedly, I'm not very familiar with the retry mechanism of the Requests module, but I don't expect it to raise a 404 exception through `raise_for_status` if the retrying have eventually succeeded.

**Comment by parthshahva** [comment]
Yea I don't think these requests would ever succeed actually even on retry. Let me check today what the best way forward here is to get this dataset to 100% completion

**Comment by rdgoite** [comment]
@parthshahva reviewing the logs, I realised the 500 errors were also caused by the upload service API and not the data store:

```
requests.exceptions.RetryError: HTTPSConnectionPool(host='upload.data.humancellatlas.org', port=443): Max retries exceeded with url: /v1/area/d3d2f98c-9b8e-4ddb-8534-2d437aa15379/process_8bfbc2bc-e48f-49ca-ba2c-2045c240954b.json (Caused by ResponseError('too many 500 error responses'))
```

This is where it we're seeing the exporter actually doing retries but eventually reaching the limit.

**Comment by rdgoite** [comment]
Please disregard my [original report on the problematic bundle files](https://github.com/HumanCellAtlas/ingest-central/issues/343#issuecomment-471973071) as those were based on an unreliable Kibana query. @aaclan-ebi has determined all the 349 bundle export attempts that failed. There were 99,495 bundles in the TM submission that were successfully exported out of 99,840 that were expected. However, we've discovered that out of the bundles that were successfully exported, 4 were duplicates and so the actual number of unique TM bundles in the DSS right now is only 99,491 and all the 349 that were reported in the logs actually failed.

Logs and related report on this can be found in [this package](https://drive.google.com/file/d/1qb9F7WSd-MWDnMe0lPgclz8bwYBU5yCJ/view?usp=sharing).

In order to complete the TM submission, we need to do the following:
* Reattempt to export the all the bundles that failed to export. The list of sequencing process id is in the `missing_process_ids.lst` file.
* Resolve the issue on duplicate bundles by removing one of the duplicates. In the `duplicate_bundles.lst` file, The bundle UUIDs are group based on the UUID of the sequencing process they are based on.

**Comment by rdgoite** [comment]
I have documented the process of tracking duplicate bundles in [this wiki page](https://github.com/HumanCellAtlas/ingest-central/wiki/Tracking-Duplicate-Bundles).

**Comment by justincc** [comment]
From my conversation with @rdgoite I believe we are now waiting on upload to give the okay before we try submitting the remaining bundles.

**Comment by malloryfreeberg** [comment]
@cjvillar (so you know the status of this)

**Comment by justincc** [comment]
@rdgoite As an aside, we as a team need to discuss how we're going to document stuff going-forward and the migration strategy. We can't have new bits in a wiki whilst other stuff is in docs and in git. #359 covers this

**Comment by rdgoite** [comment]
In the exporter, we see errors similar to the following:

```
Traceback (most recent call last):
  File "/app/src/hca-ingest/ingest/exporter/ingestexportservice.py", line 111, in export_bundle
    self.upload_metadata_files(submission_uuid, files_by_type)
  File "/app/src/hca-ingest/ingest/exporter/ingestexportservice.py", line 412, in upload_metadata_files
    raise BundleFileUploadError(message)
ingest.exporter.ingestexportservice.BundleFileUploadError: An error occurred on uploading bundle files: 404 Client Error: Not Found for url: https://upload.data.humancellatlas.org/v1/area/d3d2f98c-9b8e-4ddb-8534-2d437aa15379/process_ecc8f8b8-b655-48f5-a120-1bd282411407.json
```

Which indicates some error during the staging of files for export, as we are already aware. In my attempt to investigate this failure, I tried directly requesting for the resource that was reported to be not found in the log above. However, when I do a HTTP GET (via `curl`), I get a HTTP 500 response:

```
curl -v https://upload.data.humancellatlas.org/v1/area/d3d2f98c-9b8e-4ddb-8534-2d437aa15379/process_ecc8f8b8-b655-48f5-a120-1bd282411407.json
*   Trying 54.192.216.181...
* TCP_NODELAY set
* Connected to upload.data.humancellatlas.org (54.192.216.181) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.data.humancellatlas.org
*  start date: Feb  1 00:00:00 2019 GMT
*  expire date: Mar  1 12:00:00 2020 GMT
*  subjectAltName: host "upload.data.humancellatlas.org" matched cert's "*.data.humancellatlas.org"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fe5a8000a00)
> GET /v1/area/d3d2f98c-9b8e-4ddb-8534-2d437aa15379/process_ecc8f8b8-b655-48f5-a120-1bd282411407.json HTTP/2
> Host: upload.data.humancellatlas.org
> User-Agent: curl/7.54.0
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 500
< content-type: application/problem+json
< content-length: 583
< date: Wed, 20 Mar 2019 11:44:51 GMT
< x-amzn-requestid: 91abb61a-4b05-11e9-bed3-5f7a9c663d07
< access-control-allow-origin: *
< access-control-allow-headers: Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key
< x-amz-apigw-id: W1sj2FcWIAMFyIQ=
< x-amzn-trace-id: Root=1-5c9227b2-b94accda8b459af886463a44;Sampled=0
< x-cache: Error from cloudfront
< via: 1.1 4edba922ed0422af37c88dc3dff2ace6.cloudfront.net (CloudFront)
< x-amz-cf-id: tImaYhRQrkcAQpmRcykLQIjX9GxuCM8VVD8y8yjZXyoDzTwvPAfXFQ==
<
{
  "detail": "None is not of type 'object'\n\nFailed validating 'type' in schema['properties']['checksums']:\n    {'properties': {'crc32c': {'description': 'CRC32C', 'type': 'string'},\n                    's3_etag': {'description': 'S3 ETag', 'type': 'string'},\n                    'sha1': {'description': 'SHA1', 'type': 'string'},\n                    'sha256': {'description': 'SHA256', 'type': 'string'}},\n     'type': 'object'}\n\nOn instance['checksums']:\n    None",
  "status": 500,
  "title": "Response body does not conform to specification",
  "type": "about:blank"
}
* Connection #0 to host upload.data.humancellatlas.org left intact
```

**Comment by parthshahva** [comment]
@rdgoite apologies. I gave the green light too early. I've now confirmed that all the metadata files in the staging area have tags and no 500s should be seen. Let me know if you have any other issues when trying to re-export

**Comment by aaclan-ebi** [comment]
Thanks @parthshahva ! 

I'll proceed with retrying the failed bundles. Noting here the 349 failed exporter messages.

[https://app.zenhub.com/files/132741306/688792ad-83bb-4f5b-866c-69695609863d/download](https://app.zenhub.com/files/132741306/688792ad-83bb-4f5b-866c-69695609863d/download)

**Comment by aaclan-ebi** [comment]
349 bundles successfully reexported!

Filed ticket to DSS to delete the 4 duplicate bundles https://app.zenhub.com/workspaces/dcp-backlogs-5ac7bcf9465cb172b77760d9/issues/humancellatlas/data-store/1986

**Comment by rdgoite** [comment]
A [reference script for tombstoning duplicate bundles in the DSS](https://gist.github.com/xbrianh/39bb7ec7f4fb095ef7ffd4d2e9c5887d) was provided by blue box (see the link).

**Comment by aaclan-ebi** [comment]
Deleting the 4 bundle manifests in Ingest:

```
80ccaf4e-5284-4a94-8cd4-bba8975b1dd8:2019-02-27T210613.175220Z
ca96fdee-ec34-410b-a4ec-fcd2e45ea2aa:2019-02-27T124054.618798Z
7db2a517-4a2a-40f2-bfcf-4ff3164472ff:2019-02-27T131454.896835Z
fba487be-4c3b-4d52-bb4b-311f4715ab26:2019-02-27T023445.174104Z
```

https://api.ingest.data.humancellatlas.org/bundleManifests/search/findByBundleUuid?uuid=80ccaf4e-5284-4a94-8cd4-bba8975b1dd8
https://api.ingest.data.humancellatlas.org/bundleManifests/5c76fbc5cefdf1000730ad74

https://api.ingest.data.humancellatlas.org/bundleManifests/search/findByBundleUuid?uuid=ca96fdee-ec34-410b-a4ec-fcd2e45ea2aa
https://api.ingest.data.humancellatlas.org/bundleManifests/5c768557cefdf10007307a86

https://api.ingest.data.humancellatlas.org/bundleManifests/search/findByBundleUuid?uuid=7db2a517-4a2a-40f2-bfcf-4ff3164472ff
https://api.ingest.data.humancellatlas.org/bundleManifests/5c768d4fcefdf10007307deb

https://api.ingest.data.humancellatlas.org/bundleManifests/search/findByBundleUuid?uuid=fba487be-4c3b-4d52-bb4b-311f4715ab26
https://api.ingest.data.humancellatlas.org/bundleManifests/5c75f745cefdf10007303ff8

**Comment by aaclan-ebi** [comment]
It looks like all bundles are not indexed. See issue: https://github.com/HumanCellAtlas/data-store/issues/1990

We need to recreate all bundles to be indexed properly @rolando-ebi is working on this.
We'll be using same bundle uuids but with new versions. We're keeping track of the old versions which needs to be deleted after.

**Comment by justincc** [comment]
Bundles were re-created. We now need to assemble a list of all the bundle uuids of the old versions since datastore won't do this for us.

**Comment by justincc** [comment]
@rodrey said yesterday that datastore told us this wouldn't be necessary since the indexed ones should be ignored. However it sounds from @hannes [that this is not the case](https://humancellatlas.slack.com/archives/G9XD6L0AD/p1554276597066700?thread_ts=1553790700.002600&cid=G9XD6L0AD). @rolando-ebi when you're back, what's the status of this?
https://github.com/HumanCellAtlas/dcp/issues/320

**Comment by justincc** [comment]
Okay, so to summarize my current understanding

1. During this process we resubmitted bundles for half the dataset with indexed=true. This failed because our script had it's auth credentials revoked halfway through

2. In the initial run, we submitted a whole dataset worth of bundles with indexed=false.

The current plan is to tombstone the bundles from step 1. However, we had no plan to request tombstones of step 2, since in theory the bundles shouldn't appear for consumers

However, @hannes-ucsc [from what you said](https://humancellatlas.slack.com/archives/G9XD6L0AD/p1554276597066700?thread_ts=1553790700.002600&cid=G9XD6L0AD) it sounds like the only option you have is to receive all bundles, where you effectively ignore indexed=false? I'm presuming this blows out the reindexing time. Please could you comment/explain further.

**Comment by justincc** [comment]
@hannes-ucsc thinks [we should be easily able to find the bad bundles using the DSS search/ endpoint](https://humancellatlas.slack.com/archives/G9XD6L0AD/p1554491203133700?thread_ts=1553790700.002600&cid=G9XD6L0AD)

**Comment by rolando-ebi** [comment]
@justincc Yeah I have a script to find these old bundle versions. I didn't prioritize because I wasn't aware that this would be a big problem for the data browser. Will run it today and collect the old bundle fqids to be tombstoned/deleted

**Comment by justincc** [comment]
Rolando has now sent the list of all the bundles that need deletion (~300K) to the datastore for tombstoning.
